### PR TITLE
TS Strict update react-aria 'b' and 'L'

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-aria/listbox/src/useOption.ts [hunk 1]
```diff
@@ -117,7 +117,8 @@ export function useOption<T>(props: AriaOptionProps, state: ListState<T>, ref: R
   }
 
   if (isVirtualized) {
-    optionProps['aria-posinset'] = state.collection.getItem(key).index + 1;
+    let index = state.collection.getItem(key).index;
+    optionProps['aria-posinset'] = Number.isNaN(index) ? undefined : (index as number) + 1;
```
Comment (LFDanLu) [inline packages/@react-aria/listbox/src/useOption.ts:121]: Nit:
Doesn't `Number.isNaN` only return true if `index` is strictly `NaN`? Shouldn't we check that index is not undef as well since `index` is a optional property on a `Node`?

**Review by LFDanLu — COMMENTED** [review]
Just one nit, but otherwise looks fine to me

File: packages/@react-aria/listbox/src/useOption.ts (hunk 1 above)
Comment (snowystinger) [reply packages/@react-aria/listbox/src/useOption.ts:121]: yep, you are correct, i'll update it

File: packages/@react-aria/landmark/src/useLandmark.ts [hunk 2]
```diff
@@ -44,11 +44,11 @@ interface LandmarkManagerApi {
 // from an older version of useLandmark against a newer version of
 // LandmarkManager does not crash.
 interface Landmark {
-  ref: MutableRefObject<Element>,
+  ref: RefObject<Element>,
   role: AriaLandmarkRole,
   label?: string,
   lastFocused?: FocusableElement,
-  focus: (direction: 'forward' | 'backward') => void,
+  focus?: (direction: 'forward' | 'backward') => void,
```
Comment (devongovett) [inline packages/@react-aria/landmark/src/useLandmark.ts:51]: Why is this optional now?

File: packages/@react-aria/landmark/src/useLandmark.ts [hunk 3]
```diff
@@ -322,7 +322,7 @@ class LandmarkManager implements LandmarkManagerApi {
   private focusMain() {
     let main = this.getLandmarkByRole('main');
     if (main && document.contains(main.ref.current)) {
-      this.focusLandmark(main.ref.current, 'forward');
+      this.focusLandmark(main.ref.current as HTMLElement, 'forward');
```
Comment (devongovett) [inline packages/@react-aria/landmark/src/useLandmark.ts:325]: Maybe we should make the ref have type `RefObject<FocusableElement>`?

File: packages/@react-aria/landmark/src/useLandmark.ts [hunk 4]
```diff
@@ -391,26 +391,43 @@ class LandmarkManager implements LandmarkManagerApi {
...
   public createLandmarkController(): LandmarkController {
-    let instance = this;
+    let instance: LandmarkManager | null = this;
     instance.refCount++;
     instance.setupIfNeeded();
     return {
       navigate(direction, opts) {
-        return instance.navigate(opts?.from || document.activeElement, direction === 'backward');
+        let element = opts?.from || document?.activeElement;
+        if (element && instance) {
```
Comment (devongovett) [inline packages/@react-aria/landmark/src/useLandmark.ts:400]: I do think throwing if `instance` is null would be appropriate here since you can't use a controller after it is disposed. And seems reasonable to assert that document is defined when using these APIs as well rather than silently doing nothing. You could do `instance!.navigate(document!.activeElement, ...)` etc

File: packages/@react-aria/landmark/src/useLandmark.ts [hunk 5]
```diff
@@ -429,33 +446,33 @@ class LandmarkManager implements LandmarkManagerApi {
...
+    controller?.dispose();
     instance = getLandmarkManager();
     controller = instance.createLandmarkController();
   });
 
   // Return a wrapper that proxies requests to the current controller instance.
   return {
     navigate(direction, opts) {
-      return controller.navigate(direction, opts);
+      return controller?.navigate(direction, opts) ?? false;
```
Comment (devongovett) [inline packages/@react-aria/landmark/src/useLandmark.ts:463]: same here

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx [hunk 6]
```diff
@@ -337,14 +343,14 @@ function IframeExample() {
       });
 
       setTimeout(() => {
-        document.body.removeAttribute('data-react-aria-top-layer');
+        document?.body.removeAttribute('data-react-aria-top-layer');
       }, 100);
-    });
+    }) as EventListener);
```
Comment (devongovett) [inline packages/@react-aria/landmark/stories/Landmark.stories.tsx:348]: why?

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx [hunk 7]
```diff
@@ -337,14 +343,14 @@ function IframeExample() {
...
 
     // When the iframe is re-focused, restore focus back inside where it was before.
     window.addEventListener('focus', () => {
       if (prevFocusedElement) {
-        prevFocusedElement.focus();
+        prevFocusedElement?.focus();
```
Comment (devongovett) [inline packages/@react-aria/landmark/stories/Landmark.stories.tsx:353]: we know it's not null here because of the if statement

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx [hunk 8]
```diff
@@ -337,14 +343,14 @@ function IframeExample() {
       });
 
       setTimeout(() => {
-        document.body.removeAttribute('data-react-aria-top-layer');
+        document?.body.removeAttribute('data-react-aria-top-layer');
```
Comment (devongovett) [inline packages/@react-aria/landmark/stories/Landmark.stories.tsx:346]: ok it's going to be extremely annoying if we have to assert document is defined every time we use it 😂 

is there a tsconfig setting for that?

File: packages/@react-aria/landmark/test/useLandmark.test.tsx [hunk 9]
```diff
@@ -130,13 +131,13 @@ describe('LandmarkManager', function () {
     let navigation = tree.getByRole('navigation');
     let main = tree.getByRole('main');
 
-    fireEvent.keyDown(document.activeElement, {key: 'F6'});
-    fireEvent.keyUp(document.activeElement, {key: 'F6'});
-    expect(document.activeElement).toBe(navigation);
+    fireEvent.keyDown(document.activeElement as HTMLElement, {key: 'F6'});
```
Comment (devongovett) [inline packages/@react-aria/landmark/test/useLandmark.test.tsx:134]: why is that needed? `fireEvent` should accept any `Element` as the target.

File: packages/@react-aria/link/src/useLink.ts [hunk 10]
```diff
@@ -70,7 +70,7 @@ export function useLink(props: AriaLinkOptions, ref: RefObject<FocusableElement>
       ...linkProps,
       'aria-disabled': isDisabled || undefined,
       onClick: (e) => {
-        pressProps.onClick(e);
+        pressProps?.onClick?.(e);
```
Comment (devongovett) [inline packages/@react-aria/link/src/useLink.ts:73]: How would pressProps be null?

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx (hunk 7 above)
Comment (snowystinger) [reply packages/@react-aria/landmark/stories/Landmark.stories.tsx:353]: some of these may have been missed during a merge since I was adding strict while you were changing these files

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx (hunk 8 above)
Comment (snowystinger) [reply packages/@react-aria/landmark/stories/Landmark.stories.tsx:346]: will look, 😆

Comment (snowystinger) [reply packages/@react-aria/landmark/stories/Landmark.stories.tsx:346]: don't we need this for SSR though?

File: packages/@react-aria/landmark/test/useLandmark.test.tsx (hunk 9 above)
Comment (snowystinger) [reply packages/@react-aria/landmark/test/useLandmark.test.tsx:134]: document.activeElement is defined as `Element | null`
and fireEvent takes `Element | Node | Document | Window`
:(

I could do something like `document.activeElement!` instead though?

File: packages/@react-aria/link/src/useLink.ts (hunk 10 above)
Comment (snowystinger) [reply packages/@react-aria/link/src/useLink.ts:73]: got too excited with optional chaining

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx (hunk 8 above)
Comment (devongovett) [reply packages/@react-aria/landmark/stories/Landmark.stories.tsx:346]: I hope no one is triggering landmark navigation during SSR. That seems like user error

File: packages/@react-aria/landmark/test/useLandmark.test.tsx (hunk 9 above)
Comment (devongovett) [reply packages/@react-aria/landmark/test/useLandmark.test.tsx:134]: Ugh, why is document.activeElement defined to include null? I don't think that can ever happen in a real browser - it's always `document.body` or some other element right?

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx (hunk 8 above)
Comment (snowystinger) [reply packages/@react-aria/landmark/stories/Landmark.stories.tsx:346]: would it be better to use `!` then?
we wouldn't want to make this ok everywhere, there may be legitimate places where checking this is a good idea

File: packages/@react-aria/landmark/test/useLandmark.test.tsx (hunk 9 above)
Comment (snowystinger) [reply packages/@react-aria/landmark/test/useLandmark.test.tsx:134]: https://developer.mozilla.org/en-US/docs/Web/API/Document/activeElement#value

Comment (snowystinger) [reply packages/@react-aria/landmark/test/useLandmark.test.tsx:134]: https://html.spec.whatwg.org/multipage/interaction.html#dom-documentorshadowroot-activeelement-dev

Comment (snowystinger) [reply packages/@react-aria/landmark/test/useLandmark.test.tsx:134]: i don't know when null would ever be returned... but there is the reason for the typings

Comment (devongovett) [reply packages/@react-aria/landmark/test/useLandmark.test.tsx:134]: so according to the steps there it could occur if there is both no `<body>` or `<html>` element which seems very unlikely haha

I guess `document.activeElement!` is preferable over a type cast but still annoying. that will affect basically every single test file for an extremely edge case.

**Review by ktabors — APPROVED** [review]

File: packages/@react-aria/landmark/stories/Landmark.stories.tsx [hunk 11]
```diff
@@ -318,12 +318,18 @@ function IframeExample() {
...
+      return;
+    }
 
-    let prevFocusedElement = null;
-    window.addEventListener('react-aria-landmark-navigation', (e: CustomEvent) => {
+    let prevFocusedElement: HTMLElement | null = null;
+    window.addEventListener('react-aria-landmark-navigation', ((e: CustomEvent) => {
       e.preventDefault();
-      let el = document.activeElement;
+      if (!window || !document) {
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/landmark/stories/Landmark.stories.tsx:329]: different scope and async compared to the check above
they probably can't ever change, but TS doesn't know that
also, this is a story, so I'm not too fussed about the TS in here

**Review by ktabors — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]
LGTM

**Review by devongovett — DISMISSED** [review]

File: packages/@react-aria/landmark/src/useLandmark.ts [hunk 12]
```diff
@@ -244,10 +244,10 @@ class LandmarkManager implements LandmarkManagerApi {
   private closestLandmark(element: Element) {
     let landmarkMap = new Map(this.landmarks.map(l => [l.ref.current, l]));
     let currentElement = element;
-    while (currentElement && !landmarkMap.has(currentElement) && currentElement !== document.body) {
+    while (currentElement && !landmarkMap.has(currentElement as FocusableElement) && currentElement !== document.body && currentElement.parentElement) {
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/landmark/src/useLandmark.ts:247]: Sure will make a followup

[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/landmark/src/useLandmark.ts:247]: urgh, merge conflicts, I'll take care of it here since I'll have to get a new review anyways

**Review by devongovett — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]
LGTM
